spring-integration-core/src/main/java/org/springframework/integration/support/MessageBuilder.java

```diff
@@ -176,7 +181,16 @@ public T getPayload() {
 	 */
 	@Override
 	public MessageBuilder<T> copyHeadersIfAbsent(Map<String, ?> headersToCopy) {
-		this.headerAccessor.copyHeadersIfAbsent(headersToCopy);
+		if (headersToCopy != null) {
+			for (Map.Entry<String, ?> entry : headersToCopy.entrySet()) {
+				if (!this.headerAccessor.isReadOnly(entry.getKey())) {
+					this.headerAccessor.setHeaderIfAbsent(entry.getKey(), entry.getValue());
+				}
+				else if (logger.isWarnEnabled()) {
+					logger.warn("The header [" + entry + "] is ignored for population because it is is readOnly.");
```

Comment:
I don't think we want a WARN log for user-specified readOnlyHeaders on every build of a message.

Perhaps it would be better to put the WARN log in the for loop of `GatewayMethodInboundMessageMapper.DefaultMethodArgsMessageMapper.toMessage` Line 309 and don't invoke `headers.put(headerName, argumentValue);`